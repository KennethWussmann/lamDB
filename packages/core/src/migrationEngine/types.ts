import { z } from 'zod';

const rpcPrint = z.object({
  id: z.number(),
  method: z.literal('print'),
  params: z.object({
    content: z.string(),
  }),
});

const rpcMigrationResult = z.object({
  id: z.number(),
  result: z.object({
    appliedMigrationNames: z.array(z.string()),
  }),
});

const rpcError = z.object({
  error: z.object({
    code: z.number(),
    message: z.string(),
    data: z.object({
      is_panic: z.boolean(),
      message: z.string(),
      error_code: z.string(),
      meta: z.unknown().optional(),
    }),
  }),
});

export const rpcResponse = z.union([rpcPrint, rpcMigrationResult, rpcError]);

type ApplyMigrationRequestParams = {
  migrationsDirectoryPath: string;
};

type RPCPayload<T> = {
  id: number;
  jsonrpc: string;
  method: string;
  params: T;
};

export type ApplyMigrationRequest = RPCPayload<ApplyMigrationRequestParams>;
export type RPCResponse = z.infer<typeof rpcResponse>;
export type RPCError = z.infer<typeof rpcError>;
export type RPCMigrationResult = z.infer<typeof rpcMigrationResult>;

export const isRPCError = (rpcResponse: RPCResponse): rpcResponse is RPCError => !!(rpcResponse as RPCError).error;
export const isRPCMigrationResult = (rpcResponse: RPCResponse): rpcResponse is RPCMigrationResult =>
  !!(rpcResponse as RPCMigrationResult).result?.appliedMigrationNames;
