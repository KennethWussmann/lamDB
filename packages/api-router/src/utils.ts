import { isExecutableDefinitionNode, OperationDefinitionNode, OperationTypeNode, parse } from 'graphql';
import { Request, Response, sha1Hash } from '@lamdb/core';
import { Request as ExpressRequest, Response as ExpressResponse } from 'express';

export const graphQlErrorResponse = (message: string): Response => ({
  status: 400,
  body: JSON.stringify({
    data: null,
    errors: [
      {
        message,
      },
    ],
  }),
  headers: {
    'content-type': 'application/json',
  },
});

export const fromExpressRequest = (expressRequest: ExpressRequest): Request => ({
  method: expressRequest.method,
  body: expressRequest.body,
  headers: Object.fromEntries(
    Object.entries(expressRequest.headers).map(([key, value]) => [key, Array.isArray(value) ? value[0] : value ?? '']),
  ),
  path: expressRequest.path,
});

export const applyToExpressResponse = (response: Response, expressResponse: ExpressResponse) => {
  Object.entries(response.headers).forEach(([key, value]) => expressResponse.setHeader(key, value));
  expressResponse.status(response.status);
  expressResponse.send(response.body);
};

export const getOperationInfo = (
  request: Request,
): { name: string; type: OperationTypeNode; hash: string | undefined } | undefined => {
  if (!request.body) {
    return undefined;
  }

  let parsedDocument;
  try {
    parsedDocument = JSON.parse(request.body);
  } catch {
    return undefined;
  }

  const document = parse(parsedDocument.query, {
    noLocation: true,
  });

  const executableNodes = document.definitions.filter(isExecutableDefinitionNode);

  if (executableNodes.length === 0) {
    throw new Error('Document does not include executable nodes');
  }

  const operationNodes: OperationDefinitionNode[] = executableNodes
    .filter((node: any) => !!node?.operation)
    .map((node) => node as OperationDefinitionNode);

  if (operationNodes.length === 0) {
    throw new Error('Document does not include operation nodes');
  }
  if (operationNodes.length > 1) {
    throw new Error('Document includes multiple operation nodes: Only one is supported');
  }
  return {
    name: parsedDocument.operationName ?? operationNodes[0].name ?? '',
    type: operationNodes[0].operation,
    hash: request.body ? sha1Hash(request.body) : undefined,
  };
};
