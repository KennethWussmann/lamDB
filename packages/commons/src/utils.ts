import { createHash } from 'crypto';
import { access } from 'fs/promises';
import { Request, Response } from './requestResponse';
import { isExecutableDefinitionNode, OperationDefinitionNode, OperationTypeNode, parse } from 'graphql';

export const sha1Hash = (operation: string): string => {
  const hashsum = createHash('sha1');
  hashsum.update(operation);
  return hashsum.digest('hex').toString();
};

export const exists = async (file: string) => {
  try {
    await access(file);
    return true;
  } catch {
    return false;
  }
};

export const errorLog = (e: unknown): object => {
  const error = e as Error;
  return {
    error: {
      name: error.name,
      message: error.message,
      cause: error.cause,
      stack: error.stack,
    },
  };
};

export const graphQlErrorResponse = (message: string): Response => ({
  status: 400,
  body: JSON.stringify({
    data: null,
    errors: [
      {
        message,
      },
    ],
  }),
  headers: {
    'content-type': 'application/json',
  },
});

export const getOperationInfo = (request: Request): { name: string; type: OperationTypeNode; hash: string | undefined } | undefined => {
  if (!request.body) {
    return undefined;
  }

  let parsedDocument;
  try {
    parsedDocument = JSON.parse(request.body);
  } catch {
    return undefined;
  }

  const document = parse(parsedDocument.query, {
    noLocation: true,
  });

  const executableNodes = document.definitions.filter(isExecutableDefinitionNode);

  if (executableNodes.length === 0) {
    throw new Error('Document does not include executable nodes');
  }

  const operationNodes: OperationDefinitionNode[] = executableNodes.filter((node: any) => !!node?.operation).map((node) => node as OperationDefinitionNode);

  if (operationNodes.length === 0) {
    throw new Error('Document does not include operation nodes');
  }
  if (operationNodes.length > 1) {
    throw new Error('Document includes multiple operation nodes: Only one is supported');
  }
  return {
    name: parsedDocument.operationName ?? operationNodes[0].name ?? '',
    type: operationNodes[0].operation,
    hash: request.body ? sha1Hash(request.body) : undefined,
  };
};
