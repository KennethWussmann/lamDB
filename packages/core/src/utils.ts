import { createHash } from 'crypto';
import { access } from 'fs/promises';

export const sha1Hash = (operation: string): string => {
  const hashsum = createHash('sha1');
  hashsum.update(operation);
  return hashsum.digest('hex').toString();
};

export const exists = async (file: string) => {
  try {
    await access(file);
    return true;
  } catch {
    return false;
  }
};

export const errorLog = (e: unknown): object => {
  const error = e as Error;
  return {
    error: {
      name: error.name,
      message: error.message,
      cause: error.cause,
      stack: error.stack,
    },
  };
};

export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=5`;
