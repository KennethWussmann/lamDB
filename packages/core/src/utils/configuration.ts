// istanbul ignore file
import { config } from 'dotenv';

config();
export class LamDBConfiguration {
  get migrationEngineBinaryPath() {
    return assertEnv('MIGRATION_ENGINE_BINARY_PATH');
  }
  get migrationEngineForceMigration() {
    return assertEnv('MIGRATION_ENGINE_FORCE_MIGRATION', 'false').toLowerCase() === 'true';
  }
  get queryEngineLibraryPath() {
    return assertEnv('QUERY_ENGINE_LIBRARY_PATH');
  }
  get databasePath() {
    return assertEnv('DATABASE_PATH');
  }
  get prismaSchemaPath() {
    return assertEnv('PRISMA_SCHEMA_PATH');
  }
  get operationOptimization() {
    return assertEnv('OPERATION_OPTIMIZATION', 'false').toLowerCase() === 'true';
  }

  get server() {
    return new LamDBServerConfiguration();
  }
}

export class LamDBServerConfiguration {
  get allowCors() {
    return assertEnv('ALLOW_CORS', 'false').toLowerCase() === 'true';
  }
  get proxyPort() {
    return parseInt(assertEnv('PROXY_PORT', '4000'), 10);
  }
}

const assertEnv = (name: string, defaultValue: string | undefined = undefined): string => {
  const value = process.env[name] ?? defaultValue;
  if (!value) {
    throw new Error(`Environment variable ${name} not set, but required`);
  }
  return value;
};
