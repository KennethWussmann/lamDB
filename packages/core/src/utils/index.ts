export * from './utils';
export * from './configuration';
