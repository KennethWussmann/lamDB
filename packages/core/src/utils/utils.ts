export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=3&connection_limit=1`;
