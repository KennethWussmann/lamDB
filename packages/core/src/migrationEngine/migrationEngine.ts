import { readFile, rm, writeFile } from 'fs/promises';
import { basename, dirname, join } from 'path';
import { createLogger, logTraceSync, tracer, errorLog, exists, getDatabaseUrl, sha1Hash } from '../utils';
import execa from 'execa';
import { ApplyMigrationRequest, isRPCError, isRPCMigrationResult, RPCResponse, rpcResponse } from './types';

const logger = createLogger({ name: 'MigrationEngine' });

type MigrationEngineConfig = {
  binaryPath: string;
  prismaSchemaPath: string;
  databaseFilePath: string;
};

export class MigrationEngine {
  private messageId = 0;
  private migrationDone = false;
  private schemaHash: string | undefined;
  private previousSchemaHash: string | undefined;
  private schemaContent: string | undefined;
  private migrationLockFilePath = join(
    dirname(this.config.databaseFilePath),
    `${basename(this.config.databaseFilePath)}.migration.lock`,
  );
  private migrationsPath = join(dirname(this.config.prismaSchemaPath), 'migrations');

  constructor(private config: MigrationEngineConfig) {}

  @tracer.captureMethod({ captureResponse: false })
  private async getSchemaContent() {
    if (this.schemaContent) {
      return this.schemaContent;
    }
    this.schemaContent = await readFile(this.config.prismaSchemaPath, 'utf8');
    return this.schemaContent;
  }

  @tracer.captureMethod()
  private async getSchemaHash() {
    if (this.schemaHash) {
      return this.schemaHash;
    }
    this.schemaHash = sha1Hash(await this.getSchemaContent());
    return this.schemaHash;
  }

  @tracer.captureMethod()
  private async getPreviousSchemaHash() {
    if (this.previousSchemaHash) {
      return this.previousSchemaHash;
    }
    this.previousSchemaHash = await readFile(this.migrationLockFilePath, 'utf8');
    return this.previousSchemaHash;
  }

  @tracer.captureMethod()
  private async requiresMigration() {
    if (this.migrationDone) {
      return false;
    }
    if (!(await exists(this.migrationsPath))) {
      logger.warn('No migrations exist: Database may remain empty.');
      return false;
    }
    if (!(await exists(this.migrationLockFilePath))) {
      return true;
    }
    return (await this.getPreviousSchemaHash()) !== (await this.getSchemaHash());
  }

  @tracer.captureMethod()
  private async updateMigrationLockFile() {
    await writeFile(this.migrationLockFilePath, await this.getSchemaHash(), 'utf8');
  }

  @tracer.captureMethod()
  private async rollbackMigrationLockFile() {
    await writeFile(this.migrationLockFilePath, await this.getPreviousSchemaHash(), 'utf8');
  }

  /**
   * Check if migration is necessary using a lockfile. If lockfile is in sync with current schema, nothing will be applied.
   * If lockfile is out of sync with schema, the schema will be tried to apply using Prisma migration engine. Migration engine might still decide that nothing needs execution.
   * @param force Skip checksum lockfile check and try to apply anyways
   * @returns array of applied migration names
   */
  @tracer.captureMethod()
  async apply(force = false): Promise<string[]> {
    if (!force && !(await this.requiresMigration())) {
      logger.debug('Not running migration: Database already up-to-date');
      return [];
    }
    // optimistically lock the migration, to avoid another parallel migraiton
    await this.updateMigrationLockFile();

    logger.info('Executing migration', { schemaHash: this.schemaHash });

    let appliedMigrationNames: string[] | undefined;
    let migration: execa.ExecaChildProcess<string> | undefined;
    try {
      // Not using @prisma/migrate here because the published package is broken a lot and not accessible in a typesafe way.
      // Also it requires WASM dependencies where the path couldn't be specified.
      migration = execa(this.config.binaryPath, ['--datamodel', this.config.prismaSchemaPath], {
        env: {
          ...process.env,
          DATABASE_URL: getDatabaseUrl(this.config.databaseFilePath),
          RUST_LOG: process.env.LOG_LEVEL ?? 'info',
        },
      });
      const migrationRequest: ApplyMigrationRequest = {
        id: this.messageId++,
        jsonrpc: '2.0',
        method: 'applyMigrations',
        params: {
          migrationsDirectoryPath: this.migrationsPath,
        },
      };
      logger.debug('Executing migration egnine RPC request', { request: migrationRequest });
      migration.stdin?.write(`${JSON.stringify(migrationRequest)}\n`);

      migration.stdout?.on('data', (chunk) => {
        const responseString = Buffer.from(chunk).toString();
        try {
          logger.debug('Received migration response', { responseString });
          const response: RPCResponse = rpcResponse.parse(JSON.parse(responseString));

          if (isRPCError(response)) {
            logger.error(response.error.data.message, {
              code: response.error.code,
              error_code: response.error.data.error_code,
            });
            migration?.kill(1);
            return;
          }
          if (isRPCMigrationResult(response)) {
            appliedMigrationNames = response.result.appliedMigrationNames;
            if (!appliedMigrationNames || appliedMigrationNames.length === 0) {
              logger.info('No migration scripts required execution');
            } else {
              logger.info(
                `Executed ${appliedMigrationNames.length} migration${appliedMigrationNames.length !== 1 ? 's' : ''}`,
                { appliedMigrationNames },
              );
            }
            this.migrationDone = true;
            migration?.kill('SIGTERM');
            return;
          }
          if (response.method === 'print') {
            logger.info(response.params.content);
            // ACK message, very important, engine will wait for it
            migration?.stdin?.write(
              `${JSON.stringify({
                id: response.id,
                jsonrpc: '2.0',
                result: {},
              })}\n`,
            );
            return;
          }
        } catch (e) {
          logger.error('Received invalid migration response', { error: errorLog(e), response: responseString });
          migration?.kill(1);
        }
      });

      await migration;
    } catch (e: any) {
      // ignore a graceful SIGTERM stop, all good
      if (e?.signal !== 'SIGTERM') {
        await this.rollbackMigrationLockFile();
        logger.error('Failed to apply migrations', errorLog(e));
        throw new Error('Failed to apply migrations');
      }
    } finally {
      migration?.kill();
    }
    return appliedMigrationNames ?? [];
  }

  /**
   * Caution: Will delete the entire database files.
   * Reset the EFS to a clean state.
   */
  @tracer.captureMethod()
  async reset() {
    const databaseDir = dirname(this.config.databaseFilePath);
    await Promise.all(
      [
        this.config.databaseFilePath,
        this.migrationLockFilePath,
        `${databaseDir}-shm`,
        `${databaseDir}-wal`,
        `${databaseDir}-journal`,
      ].map(async (file) => {
        if (await exists(file)) {
          await rm(file);
        }
      }),
    );
    this.migrationDone = false;
  }
}

let migrationEngine: MigrationEngine | undefined;

export const getMigrationEngine = (config: MigrationEngineConfig): MigrationEngine =>
  logTraceSync({
    logger,
    segmentName: 'getMigrationEngine',
    metadata: { firstInit: !!migrationEngine },
    method: () => {
      if (migrationEngine) {
        return migrationEngine;
      }
      migrationEngine = new MigrationEngine(config);
      return migrationEngine;
    },
  });
