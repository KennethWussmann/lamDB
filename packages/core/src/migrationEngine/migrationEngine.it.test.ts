import { databaseMigrationLockPath, databasePath, migrationEnginePath, prismaSchemaPath } from '../../test/binaryPaths';
import { exists } from '../utils';
import { MigrationEngine } from './migrationEngine';

const engine = new MigrationEngine({
  binaryPath: migrationEnginePath,
  databaseFilePath: databasePath,
  prismaSchemaPath: prismaSchemaPath,
});

describe('migrationEngine', () => {
  it('applies migrations to fresh database and trying to reapply is skipped', async () => {
    const appliedMigrations = await engine.apply();
    const lockFileExists = await exists(databaseMigrationLockPath);

    expect(lockFileExists).toBeTruthy();
    expect(appliedMigrations).toMatchSnapshot();

    const retryAppliedMigrations = await engine.apply();
    expect(retryAppliedMigrations).toHaveLength(0);
  });
});
