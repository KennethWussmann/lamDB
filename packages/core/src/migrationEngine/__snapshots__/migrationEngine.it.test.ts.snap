// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrationEngine applies migrations to fresh database and trying to reapply is skipped 1`] = `[]`;
